File: frontend/types/mappers/auction.mapper.ts
import type {
  AuctionSummaryResponse,
  AuctionDetailResponse,
  AuctionBrowseItemResponse,
  CategoryCountResponse,
} from '@/types/api/auction.api'
import { SellerAuction, SellerAuctionStatus } from '@/types/ui/seller.ui'
import type { AuctionDetail, AuctionDetailSeller, AuctionImage } from '@/types/ui/auction.ui'
import type { AuctionBrowseItem, CategoryCount } from '@/types/ui/auction-browse.ui'
import { AuctionStatus } from '@/lib/design-tokens'

export function mapAuctionSummaryToSellerAuction(dto: AuctionSummaryResponse): SellerAuction {
  let status = SellerAuctionStatus.Draft;
  const dtoStatus = dto.status?.toUpperCase() || '';
  
  if (dtoStatus === 'ACTIVE') status = SellerAuctionStatus.Active;
  else if (dtoStatus === 'SCHEDULED') status = SellerAuctionStatus.Scheduled;
  else if (dtoStatus === 'DRAFT') status = SellerAuctionStatus.Draft;
  else if (dtoStatus === 'COMPLETED') status = SellerAuctionStatus.Completed;
  else if (dtoStatus === 'FAILED') status = SellerAuctionStatus.Failed;
  else if (dtoStatus === 'CANCELLED') status = SellerAuctionStatus.Cancelled;

  return {
    id: dto.id,
    title: dto.title,
    sellerId: dto.sellerId,
    categoryId: dto.categoryId,
    categoryName: dto.categoryName,
    startingPrice: dto.startingPrice,
    currentBid: dto.currentPrice,
    primaryImageUrl: dto.primaryImageUrl,
    totalBids: dto.totalBids,
    startsAt: new Date(dto.startTime),
    endsAt: new Date(dto.endTime),
    createdAt: new Date(dto.createdAt),
    status
  };
}

// Won/Lost/Outbid are UI-only states derived from current-user context after mapping.
// The server only returns ACTIVE, SCHEDULED, COMPLETED, CANCELLED, FAILED.
function parseStatus(status: string): AuctionStatus {
  switch (status?.toUpperCase()) {
    case 'ACTIVE':    return AuctionStatus.Active
    case 'SCHEDULED': return AuctionStatus.Scheduled
    case 'COMPLETED':
    case 'CANCELLED':
    case 'FAILED':    return AuctionStatus.Closed
    default:          return AuctionStatus.Closed
  }
}

export function mapAuctionDetailResponse(dto: AuctionDetailResponse): AuctionDetail {
  const seller: AuctionDetailSeller | null = dto.seller
    ? { id: dto.seller.id, name: dto.seller.name, avatarUrl: dto.seller.avatarUrl }
    : null

  return {
    id:              dto.id,
    title:           dto.title,
    description:     dto.description,
    categoryId:      dto.category.id,
    categoryName:    dto.category.name,
    startingPrice:   dto.startingPrice,
    bidIncrement:    dto.bidIncrement,
    buyNowPrice:     dto.buyNowPrice ?? undefined,
    depositAmount:   dto.depositAmount,
    currentBid:      dto.currentPrice,
    currentWinnerId: dto.currentWinnerId,
    totalBids:       dto.totalBids,
    status:          parseStatus(dto.status),
    startsAt:        new Date(dto.startTime),
    endsAt:          new Date(dto.endTime),
    originalEndAt:   new Date(dto.originalEndTime),
    extensionCount:  dto.extensionCount,
    completedAt:     dto.completedAt ? new Date(dto.completedAt) : undefined,
    winnerId:        dto.winnerId,
    images:          dto.images.map((img): AuctionImage => ({
      id:           img.id,
      imageUrl:     img.imageUrl,
      thumbnailUrl: img.thumbnailUrl || undefined,
      displayOrder: img.displayOrder,
      isPrimary:    img.isPrimary,
    })),
    seller,
    createdAt:       new Date(dto.createdAt),
  }
}

export function mapAuctionBrowseItem(dto: AuctionBrowseItemResponse): AuctionBrowseItem {
  return {
    id:              dto.id,
    title:           dto.title,
    primaryImageUrl: dto.primaryImageUrl,
    currentPrice:    dto.currentPrice,
    totalBids:       dto.totalBids,
    endTime:         new Date(dto.endTime),
    status:          parseStatus(dto.status),
    buyNowPrice:     dto.buyNowPrice,
    categoryName:    dto.categoryName,
  }
}

export function mapCategoryCount(dto: CategoryCountResponse): CategoryCount {
  return {
    categoryId:    dto.categoryId,
    categoryName: dto.categoryName,
    slug:         dto.slug,
    count:        dto.count,
  }
}

File: backend/auction-service/src/main/java/com/bidnow/auction/mapper/AuctionMapper.java
package com.bidnow.auction.mapper;

import com.bidnow.auction.domain.entity.AuctionCategory;
import com.bidnow.auction.domain.entity.AuctionImage;
import com.bidnow.auction.domain.entity.AuctionItem;
import com.bidnow.auction.domain.entity.AuctionStatusHistory;
import com.bidnow.auction.dto.request.UpdateAuctionRequest;
import com.bidnow.auction.dto.response.AdminAuctionDetailResponse;
import com.bidnow.auction.dto.response.AdminAuctionSummaryResponse;
import com.bidnow.auction.dto.response.AuctionBrowseItem;
import com.bidnow.auction.dto.response.AuctionCategoryResponse;
import com.bidnow.auction.dto.response.AuctionDetailResponse;
import com.bidnow.auction.dto.response.AuctionImageResponse;
import com.bidnow.auction.dto.response.AuctionStatusHistoryResponse;
import com.bidnow.auction.dto.response.AuctionSummaryResponse;
import com.bidnow.auction.dto.response.SellerAuctionResponse;
import com.bidnow.common.dto.UserSummaryResponse;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.NullValuePropertyMappingStrategy;

import java.util.List;

@Mapper(componentModel = "spring", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
public abstract class AuctionMapper {

    public SellerAuctionResponse toResponse(AuctionItem item, List<AuctionImage> images) {
        if (item == null) return null;
        return SellerAuctionResponse.builder()
                .id(item.getId())
                .sellerId(item.getSellerId())
                .title(item.getTitle())
                .description(item.getDescription())
                .category(toCategory(item.getCategory()))
                .startingPrice(item.getStartingPrice())
                .bidIncrement(item.getBidIncrement())
                .buyNowPrice(item.getBuyNowPrice())
                .depositAmount(item.getDepositAmount())
                .currentPrice(item.getCurrentPrice())
                .currentWinnerId(item.getCurrentWinnerId())
                .totalBids(item.getTotalBids())
                .status(item.getStatus())
                .startTime(item.getStartTime())
                .endTime(item.getEndTime())
                .originalEndTime(item.getOriginalEndTime())
                .extensionCount(item.getExtensionCount())
                .completedAt(item.getCompletedAt())
                .winnerId(item.getWinnerId())
                .images(images.stream().map(this::toImageResponse).toList())
                .createdAt(item.getCreatedAt())
                .updatedAt(item.getUpdatedAt())
                .build();
    }

    public AuctionSummaryResponse toSummaryResponse(AuctionItem item, AuctionImage primaryImage) {
        if (item == null) return null;
        AuctionCategory category = item.getCategory();
        return AuctionSummaryResponse.builder()
                .id(item.getId())
                .sellerId(item.getSellerId())
                .title(item.getTitle())
                .categoryId(category != null ? category.getId() : null)
                .categoryName(category != null ? category.getName() : null)
                .startingPrice(item.getStartingPrice())
                .currentPrice(item.getCurrentPrice())
                .buyNowPrice(item.getBuyNowPrice())
                .status(item.getStatus())
                .startTime(item.getStartTime())
                .endTime(item.getEndTime())
                .totalBids(item.getTotalBids())
                .primaryImageUrl(primaryImage != null ? primaryImage.getImageUrl() : null)
                .createdAt(item.getCreatedAt())
                .updatedAt(item.getUpdatedAt())
                .build();
    }

    public AuctionImageResponse toImageResponse(AuctionImage image) {
        if (image == null) return null;
        return AuctionImageResponse.builder()
                .id(image.getId())
                .imageUrl(image.getImageUrl())
                .thumbnailUrl(image.getThumbnailUrl())
                .displayOrder(image.getDisplayOrder())
                .isPrimary(image.getIsPrimary())
                .uploadedAt(image.getUploadedAt())
                .build();
    }

    public AuctionCategoryResponse toCategory(AuctionCategory category) {
        if (category == null) return null;
        return AuctionCategoryResponse.builder()
                .id(category.getId())
                .name(category.getName())
                .slug(category.getSlug())
                .build();
    }

    public AuctionDetailResponse toDetailResponse(AuctionItem item, List<AuctionImage> images, UserSummaryResponse seller) {
        if (item == null) return null;
        return AuctionDetailResponse.builder()
                .id(item.getId())
                .title(item.getTitle())
                .description(item.getDescription())
                .category(toCategory(item.getCategory()))
                .startingPrice(item.getStartingPrice())
                .bidIncrement(item.getBidIncrement())
                .buyNowPrice(item.getBuyNowPrice())
                .depositAmount(item.getDepositAmount())
                .currentPrice(item.getCurrentPrice())
                .currentWinnerId(item.getCurrentWinnerId())
                .totalBids(item.getTotalBids())
                .status(item.getStatus())
                .startTime(item.getStartTime())
                .endTime(item.getEndTime())
                .originalEndTime(item.getOriginalEndTime())
                .extensionCount(item.getExtensionCount())
                .completedAt(item.getCompletedAt())
                .winnerId(item.getWinnerId())
                .images(images.stream().map(this::toImageResponse).toList())
                .seller(seller)
                .createdAt(item.getCreatedAt())
                .build();
    }

    public AuctionBrowseItem toBrowseItem(AuctionItem item, AuctionImage primaryImage) {
        if (item == null) return null;
        AuctionCategory category = item.getCategory();
        return AuctionBrowseItem.builder()
                .id(item.getId())
                .title(item.getTitle())
                .primaryImageUrl(primaryImage != null ? primaryImage.getImageUrl() : null)
                .currentPrice(item.getCurrentPrice())
                .totalBids(item.getTotalBids())
                .endTime(item.getEndTime())
                .status(item.getStatus())
                .buyNowPrice(item.getBuyNowPrice())
                .categoryName(category != null ? category.getName() : null)
                .build();
    }

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "sellerId", ignore = true)
    @Mapping(target = "category", ignore = true)
    @Mapping(target = "currentPrice", ignore = true)
    @Mapping(target = "currentWinnerId", ignore = true)
    @Mapping(target = "totalBids", ignore = true)
    @Mapping(target = "status", ignore = true)
    @Mapping(target = "originalEndTime", ignore = true)
    @Mapping(target = "extensionCount", ignore = true)
    @Mapping(target = "completedAt", ignore = true)
    @Mapping(target = "winnerId", ignore = true)
    @Mapping(target = "winnerPaidAt", ignore = true)
    @Mapping(target = "paymentDeadline", ignore = true)
    @Mapping(target = "cancellationReason", ignore = true)
    @Mapping(target = "cancelledBy", ignore = true)
    @Mapping(target = "cancelledAt", ignore = true)
    @Mapping(target = "rejectionReason", ignore = true)
    @Mapping(target = "rejectedBy", ignore = true)
    @Mapping(target = "rejectedAt", ignore = true)
    @Mapping(target = "deletedAt", ignore = true)
    public abstract void updateFromRequest(UpdateAuctionRequest request, @MappingTarget AuctionItem item);

    public AdminAuctionSummaryResponse toAdminSummaryResponse(AuctionItem item, String sellerName) {
        if (item == null) return null;
        return AdminAuctionSummaryResponse.builder()
                .id(item.getId())
                .title(item.getTitle())
                .status(item.getStatus())
                .sellerId(item.getSellerId())
                .sellerName(sellerName)
                .category(toCategory(item.getCategory()))
                .currentPrice(item.getCurrentPrice())
                .startTime(item.getStartTime())
                .endTime(item.getEndTime())
                .totalBids(item.getTotalBids())
                .createdAt(item.getCreatedAt())
                .build();
    }

    public AdminAuctionDetailResponse toAdminDetailResponse(AuctionItem item, List<AuctionImage> images,
                                                            String sellerName, List<AuctionStatusHistoryResponse> statusHistory) {
        if (item == null) return null;
        return AdminAuctionDetailResponse.builder()
                .id(item.getId())
                .title(item.getTitle())
                .description(item.getDescription())
                .category(toCategory(item.getCategory()))
                .sellerId(item.getSellerId())
                .sellerName(sellerName)
                .startingPrice(item.getStartingPrice())
                .bidIncrement(item.getBidIncrement())
                .buyNowPrice(item.getBuyNowPrice())
                .depositAmount(item.getDepositAmount())
                .currentPrice(item.getCurrentPrice())
                .currentWinnerId(item.getCurrentWinnerId())
                .totalBids(item.getTotalBids())
                .status(item.getStatus())
                .startTime(item.getStartTime())
                .endTime(item.getEndTime())
                .originalEndTime(item.getOriginalEndTime())
                .extensionCount(item.getExtensionCount())
                .completedAt(item.getCompletedAt())
                .winnerId(item.getWinnerId())
                .cancellationReason(item.getCancellationReason())
                .cancelledBy(item.getCancelledBy())
                .cancelledAt(item.getCancelledAt())
                .rejectionReason(item.getRejectionReason())
                .rejectedBy(item.getRejectedBy())
                .rejectedAt(item.getRejectedAt())
                .images(images.stream().map(this::toImageResponse).toList())
                .statusHistory(statusHistory)
                .createdAt(item.getCreatedAt())
                .build();
    }

    public AuctionStatusHistoryResponse toStatusHistoryResponse(AuctionStatusHistory history) {
        if (history == null) return null;
        return AuctionStatusHistoryResponse.builder()
                .fromStatus(history.getFromStatus())
                .toStatus(history.getToStatus())
                .reason(history.getReason())
                .triggeredBy(history.getTriggeredBy())
                .createdAt(history.getCreatedAt())
                .build();
    }
}
